lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -777,6 +778,200 @@ static unsigned getNestedLevelInStruct(const Type *ty) {
   return lvl;
 }
 
+/// Returns first GEP index that indexes a struct member, or 0 otherwise.
+/// Ignores initial ptr index.
+static unsigned FindFirstStructMemberIdxInGEP(GEPOperator *GEP) {
+  StructType *ST = dyn_cast<StructType>(
+    GEP->getPointerOperandType()->getPointerElementType());
+  int index = 1;
+  for (auto it = gep_type_begin(GEP), E = gep_type_end(GEP); it != E;
+       ++it, ++index) {
+    if (ST) {
+      DXASSERT(!HLMatrixType::isa(ST) && !dxilutil::IsHLSLObjectType(ST),
```

Comment:
Why this will not happen?

Reply:
Should not have GEP indexing into matrix or object struct unless there's an internal bug.